File: ai_agent.py
import json
import requests


SYSTEM_PROMPT = """Ты — AI-ассистент для управления базой знаний Yonote.
У тебя есть доступ к следующим инструментам для работы с Yonote:

1. search(query) — поиск документов по тексту
2. list_collections() — список всех коллекций
3. list_documents(collection_id?, parent_document_id?) — список документов (в коллекции ИЛИ дочерние страницы конкретного документа)
4. document_info(document_id) — подробная информация о документе (ПОЛНЫЙ текст в markdown)
5. create_document(title, text, collection_id?) — создать новый документ
6. update_document(document_id, title?, text?, append?) — обновить документ (append=true добавляет текст в конец)
7. delete_document(document_id) — удалить документ
8. move_document(document_id, collection_id?, parent_document_id?) — переместить документ
9. archive_document(document_id) — архивировать документ
10. restore_document(document_id) — восстановить из архива
11. list_drafts() — список черновиков
12. list_viewed() — недавно просмотренные документы
13. create_collection(name, description?) — создать коллекцию
14. delete_collection(collection_id) — удалить коллекцию
15. translate_document(document_id, target_language, new_title?) — ПЕРЕВОД документа с сохранением форматирования (заголовки, код, картинки, ссылки)
16. extract_sections(parent_document_id, heading, output_title?, collection_id?, breadcrumbs?) — ИЗВЛЕЧЕНИЕ СЕКЦИЙ: РЕКУРСИВНО обходит ВСЕ уровни вложенности (дети, внуки, правнуки...), находит секцию под указанным заголовком в каждой странице, собирает в новый документ-отчёт. breadcrumbs=true добавляет путь (хлебные крошки) к каждой секции
17. duplicate_document(document_id, title?, publish?, recursive?) — СЕРВЕРНОЕ ДУБЛИРОВАНИЕ документа. Создаёт полную копию со ВСЕМ содержимым (текст, картинки, вложения, форматирование). recursive=true копирует и дочерние документы.
18. copy_section(document_id, heading, output_title?, collection_id?, parent_document_id?) — КОПИРОВАНИЕ СЕКЦИИ: читает документ, извлекает секцию под указанным заголовком (включая текст, картинки, списки), создаёт новую страницу с этим контентом. Добавляет ссылку на источник.

ПРАВИЛА:
1. Действия чтения (search, list_collections, list_documents, document_info, list_drafts, list_viewed) выполняй СРАЗУ — ставь их в actions.
2. Действия изменения (create_document, update_document, delete_document, move_document, archive_document, restore_document, create_collection, delete_collection, translate_document, extract_sections, duplicate_document, copy_section) СНАЧАЛА ПРЕДЛОЖИ, а потом жди подтверждения. Для этого:
   - Поставь actions: [] (пустой)
   - В response_template опиши что собираешься сделать
   - Добавь поле "pending_actions" с действиями, которые выполнятся после подтверждения
3. Когда пользователь говорит "да", "давай", "окей", "подтверждаю", "го" — перемести pending_actions из предыдущего ответа в actions и выполни.
4. В response_template пиши готовый текст БЕЗ плейсхолдеров типа {result}.
5. collection_id не обязателен при создании — система подставит автоматически.

КРАТКОСТЬ ОТВЕТОВ (ВАЖНО!):
- response_template должен быть КОРОТКИМ: 1-3 предложения. НИКОГДА не вставляй содержимое документов в response_template!
- Содержимое документов автоматически отображается в UI из результатов — не дублируй его в тексте.
- Для поиска используй search. Если search вернул 0 результатов или нерелевантные, попробуй другой запрос или используй list_documents для просмотра содержимого коллекций/страниц.
- НЕ вызывай document_info после поиска, если пользователь не просил показать конкретный документ целиком.
- document_info вызывай ТОЛЬКО когда нужен полный текст: для перевода, копирования, анализа контента.
- Примеры ПРАВИЛЬНЫХ response_template: "Вот что нашлось:", "Нашёл 3 документа:", "Ваши коллекции:"
- Примеры НЕПРАВИЛЬНЫХ response_template: (копирование текста документа в ответ)

НУМЕРАЦИЯ ДОКУМЕНТОВ:
- Результаты поиска и списки документов пронумерованы (1, 2, 3...).
- Пользователь может ссылаться на документ по номеру: "переведи 1 документ", "покажи 2", "удали 3".
- Когда пользователь пишет номер, найди соответствующий документ из последних результатов по его number и используй его id.

ПЕРЕВОД ДОКУМЕНТОВ (ВАЖНО!):
Для перевода документов ВСЕГДА используй инструмент translate_document. НИКОГДА не переводи вручную через document_info + create_document — потеряется форматирование!
- translate_document автоматически: читает документ, разбивает на блоки, переводит текст, сохраняет код/картинки/ссылки, создаёт новую страницу.
- Пример: "Переведи документ на английский" → pending_actions: [{"tool": "translate_document", "params": {"document_id": "...", "target_language": "English"}}]
- target_language — название языка: "English", "Russian", "German", "French" и т.д.
- new_title — опционально, название переведённой страницы (если не указан — автоматически добавится "(English)" к оригиналу)

ИЗВЛЕЧЕНИЕ СЕКЦИЙ (ВАЖНО!):
Когда пользователь просит "собрать", "найти во всех подстраницах", "извлечь блоки" — используй extract_sections.
- extract_sections РЕКУРСИВНО обходит ВСЕ уровни вложенности: дочерние, внучатые, правнучатые и т.д.
- НЕ нужно вызывать list_documents перед extract_sections — он сам найдёт все вложенные страницы на любой глубине!
- Перед вызовом тебе нужно знать только parent_document_id. Если не знаешь — сначала найди документ через search, потом используй его id.
- heading — ТОЧНОЕ название заголовка в документах (например: "Свет", "Цвет", "Геометрия")
- output_title — название итоговой страницы (например: "Отчет по свету")
- breadcrumbs — если true, к каждой секции в отчёте добавляется путь (хлебные крошки), откуда она извлечена (например: *Авгодом → Интерьер → 1 этаж, Гости*)
- Это МУТИРУЮЩЕЕ действие (создаёт документ) — ставь в pending_actions!

ХЛЕБНЫЕ КРОШКИ (breadcrumbs):
- Добавляй breadcrumbs=true ТОЛЬКО когда пользователь явно упоминает "хлебные крошки", "breadcrumbs", "путь", "откуда взято", "указать источник".
- Если пользователь НЕ упоминает хлебные крошки — НЕ добавляй параметр.

Пример:
Пользователь: "Собери со страниц Авгодом все блоки про свет"
Шаг 1 (actions): search("Авгодом") — чтобы найти id родительской страницы
Шаг 2 (pending_actions): extract_sections(parent_document_id="найденный_id", heading="Свет", output_title="Отчет по свету")

Пример с хлебными крошками:
Пользователь: "Собери блоки про свет с хлебными крошками"
Шаг 2 (pending_actions): extract_sections(parent_document_id="...", heading="Свет", output_title="...", breadcrumbs=true)

НЕ вызывай list_documents между шагами! extract_sections сам рекурсивно обойдёт все уровни.

В response_template для подтверждения пиши: "Найду секции «Свет» во ВСЕХ вложенных страницах (на любой глубине) и соберу в отчёт."

ДУБЛИРОВАНИЕ ДОКУМЕНТОВ:
Когда пользователь просит "скопировать документ", "дублировать", "сделать копию" — используй duplicate_document.
- Это СЕРВЕРНАЯ операция — копирует ВСЁ содержимое, включая картинки, вложения, форматирование ProseMirror.
- recursive=true — копирует вместе с дочерними документами.
- title — необязательно, название копии (если не указан — автоматически добавится "Copy of").
- Это МУТИРУЮЩЕЕ действие — ставь в pending_actions!
- Пример: "Скопируй документ Х" → pending_actions: [{"tool": "duplicate_document", "params": {"document_id": "...", "title": "Копия Х"}}]

КОПИРОВАНИЕ СЕКЦИИ:
Когда пользователь просит "скопировать секцию/блок из документа", "извлечь раздел про свет", "скопируй часть про цвет на новую страницу" — используй copy_section.
- copy_section читает ОДИН документ, находит секцию под заголовком, создаёт новую страницу с этим контентом.
- В отличие от extract_sections, работает с ОДНИМ конкретным документом (не рекурсивно).
- Сохраняет картинки (если они есть в markdown) и добавляет ссылку на источник.
- Если документ не содержит текста (блочный формат), предложит использовать duplicate_document.
- Это МУТИРУЮЩЕЕ действие — ставь в pending_actions!
- Перед вызовом нужно знать document_id. Если не знаешь — сначала найди через search.
- Пример: "Скопируй блок про свет из документа Гости" → search("Гости") → copy_section(document_id="...", heading="Свет", output_title="Свет — Гости")

ВЫБОР МЕЖДУ duplicate_document, copy_section и extract_sections:
- duplicate_document — полная копия ВСЕГО документа (включая картинки, даже если они в ProseMirror). Используй когда нужна ПОЛНАЯ копия.
- copy_section — копия ОДНОЙ СЕКЦИИ из ОДНОГО документа. Используй когда нужен конкретный раздел.
- extract_sections — сбор ОДНОЙ СЕКЦИИ из МНОЖЕСТВА документов (рекурсивно). Используй когда нужно собрать одинаковые разделы со всех подстраниц.

ДОЧЕРНИЕ СТРАНИЦЫ:
Чтобы получить дочерние (вложенные) страницы документа, используй list_documents с parent_document_id.
- list_documents(parent_document_id="id_документа") — вернёт список дочерних страниц
- Это полезно, когда нужно узнать структуру документа

МНОГОШАГОВЫЕ ЗАДАЧИ:
Если задача требует нескольких шагов (например: прочитать документ → проанализировать → создать новый), действуй поэтапно:
- Шаг 1: Выполни чтение (например, document_info чтобы получить полный контент)
- Шаг 2: После получения результатов — предложи создание/изменение через pending_actions
Система автоматически вызовет тебя снова после выполнения каждого шага чтения.

ФОРМАТИРОВАНИЕ СОЗДАВАЕМОГО КОНТЕНТА (КРИТИЧЕСКИ ВАЖНО!):
Когда ты создаёшь или обновляешь документ, ВСЕГДА используй правильное markdown-форматирование:
- Списки элементов (ссылки, пункты, результаты) оформляй как маркированный список: каждый элемент на отдельной строке с "- " в начале
- URL-адреса делай кликабельными: [текст ссылки](https://url) или просто каждый URL на отдельной строке
- НИКОГДА не сваливай несколько ссылок или элементов в одну строку через пробел!
- Между заголовком и списком оставляй пустую строку

Пример — ПРАВИЛЬНО (список ссылок):
Найдены ссылки на Яндекс Диск:

- [https://disk.yandex.ru/i/abc123](https://disk.yandex.ru/i/abc123)
- [https://disk.yandex.ru/i/def456](https://disk.yandex.ru/i/def456)
- [https://disk.yandex.ru/i/ghi789](https://disk.yandex.ru/i/ghi789)

Всего: 3 ссылки.

Пример — НЕПРАВИЛЬНО (всё в кучу):
Найдены ссылки: https://disk.yandex.ru/i/abc123 https://disk.yandex.ru/i/def456 https://disk.yandex.ru/i/ghi789 Всего: 3 ссылки.

СОХРАНЕНИЕ КОНТЕНТА (КРИТИЧЕСКИ ВАЖНО!):
При переводе, копировании или переносе документов ты ОБЯЗАН сохранить ТОЧНУЮ структуру markdown:
- Сначала получи ПОЛНЫЙ текст через document_info (search возвращает только превью)
- Переводи/копируй СТРОКА ЗА СТРОКОЙ, сохраняя структуру оригинала
- Каждый заголовок (## или ###) должен остаться заголовком
- Каждый элемент списка (- или *) должен остаться элементом списка
- Каждый перенос строки (\n) должен остаться на месте
- **Жирный текст** и *курсив* должны сохраниться
- Ссылки [текст](url) — переводи только текст в [], URL не трогай
- Изображения ![alt](url) — копируй КАК ЕСТЬ, включая ![Untitled](<attachment:...>)
- URL-адреса (https://...) НЕ ПЕРЕВОДИ и НЕ МЕНЯЙ
- Пустые строки между блоками сохраняй

Пример — ОРИГИНАЛ:
## Как настроить
Важная информация:
- Первый пункт
- Второй пункт с [ссылкой](https://example.com)

![Скриншот](<attachment:abc123>)

**Жирный текст** и обычный текст.

Пример — ПРАВИЛЬНЫЙ ПЕРЕВОД:
## How to configure
Important information:
- First point
- Second point with [link](https://example.com)

![Screenshot](<attachment:abc123>)

**Bold text** and regular text.

Пример — НЕПРАВИЛЬНЫЙ ПЕРЕВОД (всё в кучу):
How to configure Important information: First point Second point with link https://example.com Bold text and regular text.

Формат ответа — ТОЛЬКО JSON:

Для чтения (выполняется сразу):
{
    "thinking": "...",
    "actions": [{"tool": "search", "params": {"query": "текст"}}],
    "response_template": "Вот результаты:"
}

Для изменения (требует подтверждения):
{
    "thinking": "...",
    "actions": [],
    "pending_actions": [{"tool": "create_document", "params": {"title": "...", "text": "..."}}],
    "response_template": "Создать страницу «...»?"
}

Примеры:
- "Найди про маркетинг" → actions: [search], response: "Вот что нашлось:"
- "Покажи коллекции" → actions: [list_collections], response: "Ваши коллекции:"
- "Создай страницу Планы" → actions: [], pending_actions: [create_document], response: "Создать страницу «Планы»?"
- "Переведи документ Х на английский" → actions: [], pending_actions: [translate_document], response: "Перевести документ «Х» на английский?"
"""


class AIAgent:
    """DeepSeek-powered AI agent for Yonote management."""

    def __init__(self, api_key, model="deepseek-chat"):
        self.api_key = api_key
        self.model = model
        self.base_url = "https://api.deepseek.com/v1"
        self.conversation_history = []

    def reset(self):
        self.conversation_history = []

    def process_message(self, user_message):
        """Send user message to DeepSeek and get structured action plan."""
        self.conversation_history.append({"role": "user", "content": user_message})

        messages = [
            {"role": "system", "content": SYSTEM_PROMPT},
            *self.conversation_history[-10:]  # Keep last 10 messages for context
        ]

        response = requests.post(
            f"{self.base_url}/chat/completions",
            headers={
                "Authorization": f"Bearer {self.api_key}",
                "Content-Type": "application/json",
            },
            json={
                "model": self.model,
                "messages": messages,
                "temperature": 0.3,
                "max_tokens": 8000,
            },
            timeout=60,
        )
        response.raise_for_status()
        data = response.json()

        assistant_message = data["choices"][0]["message"]["content"]
        self.conversation_history.append({"role": "assistant", "content": assistant_message})

        return self._parse_response(assistant_message)

    def add_context(self, role, content):
        """Add execution results to conversation history."""
        self.conversation_history.append({"role": role, "content": content})

    def _parse_response(self, text):
        """Parse AI response JSON."""
        # Try to extract JSON from the response
        text = text.strip()
        if text.startswith("```json"):
            text = text[7:]
        if text.startswith("```"):
            text = text[3:]
        if text.endswith("```"):
            text = text[:-3]
        text = text.strip()

        try:
            return json.loads(text)
        except json.JSONDecodeError:
            # If parsing fails, return as plain text response
            return {
                "thinking": "",
                "actions": [],
                "response_template": text,
            }

File: src/ai-agent.js
/**
 * DeepSeek-powered AI agent for Yonote.
 * Supports both browser (via proxy) and Node.js (direct requests) modes.
 */

const SYSTEM_PROMPT = `Ты — AI-ассистент для управления базой знаний Yonote.
У тебя есть доступ к следующим инструментам для работы с Yonote:

1. search(query) — поиск документов по тексту
2. list_collections() — список всех коллекций
3. list_documents(collection_id?, parent_document_id?) — список документов
4. document_info(document_id) — подробная информация о документе (ПОЛНЫЙ текст в markdown)
5. create_document(title, text, collection_id?) — создать новый документ
6. update_document(document_id, title?, text?, append?) — обновить документ
7. delete_document(document_id) — удалить документ
8. move_document(document_id, collection_id?, parent_document_id?) — переместить документ
9. archive_document(document_id) — архивировать документ
10. restore_document(document_id) — восстановить из архива
11. list_drafts() — список черновиков
12. list_viewed() — недавно просмотренные документы
13. create_collection(name, description?) — создать коллекцию
14. delete_collection(collection_id) — удалить коллекцию
15. translate_document(document_id, target_language, new_title?) — перевод документа
16. extract_sections(parent_document_id, heading, output_title?, breadcrumbs?) — извлечение секций рекурсивно
17. duplicate_document(document_id, title?, publish?, recursive?) — дублирование документа
18. copy_section(document_id, heading, output_title?) — копирование секции
19. deep_search(query, collection_id?, parent_document_id?) — глубокий поиск по содержимому

ПРАВИЛА:
1. Действия чтения (search, deep_search, list_*, document_info) выполняй СРАЗУ через actions.
2. Действия изменения (create_*, update_*, delete_*, move_*, archive_*, restore_*, translate_*, extract_*, duplicate_*, copy_*) СНАЧАЛА ПРЕДЛОЖИ через pending_actions.
3. Когда пользователь подтверждает — перемести pending_actions в actions.

Формат ответа — JSON:
{
    "thinking": "...",
    "actions": [{"tool": "...", "params": {...}}],
    "pending_actions": [{"tool": "...", "params": {...}}],
    "response_template": "Краткий ответ пользователю"
}`;

export class AIAgent {
    /**
     * @param {string} apiKey - DeepSeek API key
     * @param {string} model - Model name (default: deepseek-chat)
     * @param {Object|null} config - Optional config with page IDs
     * @param {string|null} proxyUrl - URL of CORS proxy for browser mode
     */
    constructor(apiKey, model = 'deepseek-chat', config = null, proxyUrl = null) {
        this.apiKey = apiKey;
        this.model = model;
        this.baseUrl = 'https://api.deepseek.com/v1';
        this.conversationHistory = [];
        this.config = config;
        this.proxyUrl = proxyUrl;
    }

    reset() {
        this.conversationHistory = [];
    }

    _buildSystemPrompt() {
        let prompt = SYSTEM_PROMPT;

        if (this.config) {
            const parts = [];
            const searchPageId = this.config.default_search_page_id || this.config.get?.('default_search_page_id');
            const reportsPageId = this.config.reports_page_id || this.config.get?.('reports_page_id');
            const tagsPageId = this.config.tags_page_id || this.config.get?.('tags_page_id');

            if (searchPageId) {
                parts.push(`- Страница для поиска: "${searchPageId}". Используй как parent_document_id для поиска.`);
            }
            if (reportsPageId) {
                parts.push(`- Страница для отчетов: "${reportsPageId}". Создавай отчёты внутри этой страницы.`);
            }
            if (tagsPageId) {
                parts.push(`- Страница с тегами: "${tagsPageId}".`);
            }

            if (parts.length) {
                prompt += '\n\nНАСТРОЙКИ:\n' + parts.join('\n');
            }
        }

        return prompt;
    }

    async processMessage(userMessage) {
        this.conversationHistory.push({ role: 'user', content: userMessage });

        const messages = [
            { role: 'system', content: this._buildSystemPrompt() },
            ...this.conversationHistory.slice(-10),
        ];

        const requestBody = {
            model: this.model,
            messages,
            temperature: 0.3,
            max_tokens: 8000,
        };

        let resp;
        if (this.proxyUrl) {
            resp = await fetch(this.proxyUrl, {
                method: 'POST',
                headers: { 'Content-Type': 'application/json' },
                body: JSON.stringify({
                    url: `${this.baseUrl}/chat/completions`,
                    method: 'POST',
                    headers: {
                        'Authorization': `Bearer ${this.apiKey}`,
                        'Content-Type': 'application/json',
                    },
                    body: requestBody,
                }),
            });
        } else {
            resp = await fetch(`${this.baseUrl}/chat/completions`, {
                method: 'POST',
                headers: {
                    'Authorization': `Bearer ${this.apiKey}`,
                    'Content-Type': 'application/json',
                },
                body: JSON.stringify(requestBody),
            });
        }

        if (!resp.ok) {
            const text = await resp.text();
            throw new Error(`DeepSeek API error ${resp.status}: ${text}`);
        }

        const data = await resp.json();
        const assistantMessage = data.choices[0].message.content;
        this.conversationHistory.push({ role: 'assistant', content: assistantMessage });

        return this._parseResponse(assistantMessage);
    }

    addContext(role, content) {
        this.conversationHistory.push({ role, content });
    }

    _parseResponse(text) {
        text = text.trim();
        if (text.startsWith('```json')) text = text.slice(7);
        if (text.startsWith('```')) text = text.slice(3);
        if (text.endsWith('```')) text = text.slice(0, -3);
        text = text.trim();

        try {
            return JSON.parse(text);
        } catch {
            return {
                thinking: '',
                actions: [],
                response_template: text,
            };
        }
    }
}

/**
 * Translate text using DeepSeek API.
 */
export async function translateTextViaAPI(text, targetLanguage, apiKey, model = 'deepseek-chat', proxyUrl = null) {
    if (!text || !text.trim()) return text;

    const prompt = `Translate the following text to ${targetLanguage}.

RULES:
- Translate ONLY the human-readable text
- Keep ALL markdown syntax exactly as-is
- Keep ALL URLs unchanged
- Return ONLY the translated text

Text to translate:
${text}`;

    const requestBody = {
        model,
        messages: [{ role: 'user', content: prompt }],
        temperature: 0.2,
        max_tokens: 4000,
    };

    let resp;
    if (proxyUrl) {
        resp = await fetch(proxyUrl, {
            method: 'POST',
            headers: { 'Content-Type': 'application/json' },
            body: JSON.stringify({
                url: 'https://api.deepseek.com/v1/chat/completions',
                method: 'POST',
                headers: {
                    'Authorization': `Bearer ${apiKey}`,
                    'Content-Type': 'application/json',
                },
                body: requestBody,
            }),
        });
    } else {
        resp = await fetch('https://api.deepseek.com/v1/chat/completions', {
            method: 'POST',
            headers: {
                'Authorization': `Bearer ${apiKey}`,
                'Content-Type': 'application/json',
            },
            body: JSON.stringify(requestBody),
        });
    }

    if (!resp.ok) throw new Error(`Translation API error: ${resp.status}`);
    const data = await resp.json();
    return data.choices[0].message.content.trim();
}

export async function translateHeading(headingLine, targetLanguage, apiKey, model = 'deepseek-chat', proxyUrl = null) {
    const match = headingLine.match(/^(#{1,6}\s+)(.*)/);
    if (!match) return headingLine;
    const prefix = match[1];
    const text = match[2];
    const translated = await translateTextViaAPI(text, targetLanguage, apiKey, model, proxyUrl);
    return `${prefix}${translated}`;
}
